beginner_source/ddp_multigpu.rst

```diff
@@ -0,0 +1,182 @@
+----------------------------------------------------
+
+These are the changes you typically make to a single-GPU training script to enable DDP.
+
+Imports
+~~~~~~~
+
+.. code:: diff
+
+   import torch
+   import torch.nn.functional as F
+   from utils import MyTrainDataset
+    
+   + import torch.multiprocessing as mp
+   + from torch.utils.data.distributed import DistributedSampler
+   + from torch.nn.parallel import DistributedDataParallel as DDP
+   + from torch.distributed import init_process_group, destroy_process_group
+   + import os
+
+-  ``torch.multiprocessing`` is a PyTorch wrapper around python's native
+   multiprocessing
+-  The dsitributed process group contains all the processes that can
+   communicate and synchronize with each other.
+
+Constructing the process group
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: diff
+
+   + def ddp_setup(rank, world_size):
+   +   """
+   +   Args:
+   +       rank: Unique identifier of each process
+   +      world_size: Total number of processes
+   +   """
+   +   os.environ["MASTER_ADDR"] = "localhost"
+   +   os.environ["MASTER_PORT"] = "12355"
+   +   init_process_group(backend="nccl", rank=rank, world_size=world_size)
+
+-  `Choosing a DDP
+   backend <https://pytorch.org/docs/stable/distributed.html#which-backend-to-use>`__
+-  The process group can be initialized by TCP (default) or from a
+   shared file-system. `Read more on process group
+   initialization <https://pytorch.org/docs/stable/distributed.html#tcp-initialization>`__
+-  `init_process_group <https://pytorch.org/docs/stable/distributed.html?highlight=init_process_group#torch.distributed.init_process_group>`__
+   initializes the distributed process group.
+
+Constructing the DDP model
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: diff
+
+   - self.model = model.to(gpu_id)
+   + self.model = DDP(model, device_ids=[gpu_id])
+
+Distributing input data
+~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: diff
+
+   train_data = torch.utils.data.DataLoader(
+       dataset=train_dataset,
+       batch_size=32,
+   -   shuffle=True,
+   +   shuffle=False,
+   +   sampler=DistributedSampler(train_dataset),
+   )
+
+-  `DistributedSampler <https://pytorch.org/docs/stable/data.html?highlight=distributedsampler#torch.utils.data.distributed.DistributedSampler>`__
+   chunks the input data across all distributed processes.
+-  Each process will receive an input batch of 32 samples; the effective
+   batch size is ``32 * nprocs``, or 128 when using 4 GPUs.
+
+Saving model checkpoints
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: diff
+
+   - ckp = self.model.state_dict()
+   + ckp = self.model.module.state_dict()
+   ...
+   ...
+   - if epoch % self.save_every == 0:
+   + if self.gpu_id == 0 and epoch % self.save_every == 0:
+      self._save_checkpoint(epoch)
+
+We only need to save model checkpoints from one process. Without this
```

Comment:
mention that if this is being done, `_save_checkpoint` should not have any collective calls

Reply:
Can you explain this more? Why should we not have collective calls here, and what collective calls might typically be needed while saving checkpoints?

Reply:
Any code that runs on a specific (or subset) of ranks in a group should not have collectives because collectives would require all ranks to run the collective code.

In _save_checkpoint, sometimes users might have collectives to gather certain states across a rank etc, so that wouldn't work and typically the code would have to be moved to before the `if rank == 0` check.